src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Windows.cs

```diff
@@ -27,126 +28,210 @@ public partial class Ping
+            // Since isAsync == true, DoSendPingCore will execute asynchronously and return an active Task 
+            return DoSendPingCore(address, buffer, timeout, options, isAsync: true);
+        }
+
         // Any exceptions that escape synchronously will be caught by the caller and wrapped in a PingException.
         // We do not need to or want to capture such exceptions into the returned task.
-        private Task<PingReply> SendPingAsyncCore(IPAddress address, byte[] buffer, int timeout, PingOptions options)
+        private Task<PingReply> DoSendPingCore(IPAddress address, byte[] buffer, int timeout, PingOptions options, bool isAsync)
         {
-            var tcs = new TaskCompletionSource<PingReply>();
-            _taskCompletionSource = tcs;
+            TaskCompletionSource<PingReply> tcs = null;
+            if (isAsync)
+            {
+                _taskCompletionSource = tcs = new TaskCompletionSource<PingReply>();
+            }
 
             _ipv6 = (address.AddressFamily == AddressFamily.InterNetworkV6);
             _sendSize = buffer.Length;
 
-            // Get and cache correct handle.
-            if (!_ipv6 && _handlePingV4 == null)
+            // Cache correct handle.
+            InitialiseIcmpHandle();
+
+            if (_replyBuffer == null)
             {
-                _handlePingV4 = Interop.IpHlpApi.IcmpCreateFile();
-                if (_handlePingV4.IsInvalid)
+                _replyBuffer = SafeLocalAllocHandle.LocalAlloc(MaxUdpPacket);
+            }
+
+            int error;
+            try
+            {
+                if (isAsync)
                 {
-                    _handlePingV4 = null;
-                    throw new Win32Exception(); // Gets last error.
+                    RegisterWaitHandle();
                 }
+
+                SetUnmanagedStructures(buffer);
+
+                error = SendEcho(address, buffer, timeout, options, isAsync);
             }
-            else if (_ipv6 && _handlePingV6 == null)
+            catch
             {
-                _handlePingV6 = Interop.IpHlpApi.Icmp6CreateFile();
-                if (_handlePingV6.IsInvalid)
+                Cleanup(isAsync, isComplete: false);
+                throw;
+            }
+
+            if (error == 0)
+            {
+                error = Marshal.GetLastWin32Error();
+
+                // Only skip Async IO Pending error value.
+                if (!isAsync || error != Interop.IpHlpApi.ERROR_IO_PENDING)
                 {
-                    _handlePingV6 = null;
-                    throw new Win32Exception(); // Gets last error.
+                    Cleanup(isAsync, isComplete: false);
+                    throw new Win32Exception(error);
                 }
             }
 
-            var ipOptions = new Interop.IpHlpApi.IPOptions(options);
+            if (isAsync)
+                return tcs.Task;
 
-            if (_replyBuffer == null)
+            Cleanup(isAsync: true, isComplete: true);
```

Comment:
Shouldn't isAsync be false here?

Reply:
I think you're probably right. On my phone at the minute but will check properly soon and submit update